SmartDeviceLink/SDLSoftButtonManager.m

```diff
@@ -0,0 +1,291 @@
+
+    return self;
+}
+
+- (void)setSoftButtonObjects:(NSArray<SDLSoftButtonObject *> *)softButtonObjects {
+    self.inProgressUpdate = nil;
+    if (self.inProgressHandler != nil) {
+        self.inProgressHandler([NSError sdl_softButtonManager_pendingUpdateSuperseded]);
+        self.inProgressHandler = nil;
+    }
+    self.hasQueuedUpdate = NO;
+    if (self.queuedUpdateHandler != nil) {
+        self.queuedUpdateHandler([NSError sdl_softButtonManager_pendingUpdateSuperseded]);
+        self.queuedUpdateHandler = nil;
+    }
+
+    // Set the soft button ids. Check to make sure no two soft buttons have the same name, there aren't many soft buttons, so n^2 isn't going to be bad
+    for (NSUInteger i = 0; i < softButtonObjects.count; i++) {
+        NSString *buttonName = softButtonObjects[i].name;
+        softButtonObjects[i].buttonId = i * 100;
+        for (NSUInteger j = (i + 1); j < softButtonObjects.count; j++) {
+            if ([softButtonObjects[j].name isEqualToString:buttonName]) {
+                _softButtonObjects = @[];
+                return;
+            }
+        }
+    }
+
+    _softButtonObjects = softButtonObjects;
+
+    for (SDLSoftButtonObject *button in _softButtonObjects) {
+        button.manager = self;
+    }
+
+    NSMutableArray<SDLArtwork *> *initialStatesToBeUploaded = [NSMutableArray array];
+    NSMutableArray<SDLArtwork *> *otherStatesToBeUploaded = [NSMutableArray array];
+    if (self.displayCapabilities ? self.displayCapabilities.graphicSupported.boolValue : YES) {
+        // Upload all soft button images, the initial state images first, then the other states. We need to send updates when the initial state is ready.
+        for (SDLSoftButtonObject *object in self.softButtonObjects) {
+            if (object.currentState.artwork != nil && ![self.fileManager hasUploadedFile:object.currentState.artwork]) {
+                [initialStatesToBeUploaded addObject:object.currentState.artwork];
+            }
+        }
+        for (SDLSoftButtonObject *object in self.softButtonObjects) {
+            for (SDLSoftButtonState *state in object.states) {
+                if ([state.name isEqualToString:object.currentState.name]) { continue; }
+                if (state.artwork != nil && ![self.fileManager hasUploadedFile:state.artwork]) {
+                    [otherStatesToBeUploaded addObject:state.artwork];
+                }
+            }
+        }
+    }
+
+    // Upload initial images, then other state images
+    if (initialStatesToBeUploaded.count > 0) {
+        [self.fileManager uploadArtworks:[initialStatesToBeUploaded copy] completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+            [self sdl_updateWithCompletionHandler:nil];
+        }];
+    }
+    if (otherStatesToBeUploaded.count > 0) {
```

Comment:
If I upload invalid image data, the `SDLSoftButtonManager` tries to send a soft button with `type` of `TEXT` but with no text. This `show` RPC fails with an `INVALID_DATA` result code. Is it possible to just return an error instead of sending the invalid soft button?

Reply:
We can't return an error because this is the setter. I have added error logs however. Is this sufficient? I'm not quite sure the path your code is taking, the soft button shouldn't send in `sdl_textButtonsForCurrentState` if there's no text on _any_ of the buttons, it will just wait for the images to upload.

Reply:
It looks like the error occurs the second time the request is sent.
## 1st Update
One text-field set, one soft button with an invalid image (no text set for the button). This request returns `SUCCESS`.
```
2018-03-15 13:57:45.841869-0400 SDL_Smoke_Tests[10979:1087712] Text & Graphic update request: Show (request)
{
alignment = CENTERED;
graphic = "{\n    imageType = DYNAMIC;\n    value = \"sdl_BlankArt\";\n}";
mainField1 = "button with broken image";
mainField2 = "";
mainField3 = "";
mainField4 = "";
metadataTags = "{\n    mainField1 =     (\n    );\n}";
secondaryGraphic = "{\n    imageType = DYNAMIC;\n    value = \"sdl_BlankArt\";\n}";
}, response: Show (response)
{
resultCode = SUCCESS;
success = 1;
}, error: (null)
2018-03-15 13:57:45.842027-0400 SDL_Smoke_Tests[10979:1087703] [Proxy] Message received: Show (response)
{
resultCode = SUCCESS;
success = 1;
}
```
## 2nd Update
Same text-field and button as the 1st update. The second request returns `INVALID_DATA`.
```
2018-03-15 13:58:56.184273-0400 SDL_Smoke_Tests[10979:1094326] SoftButton update request: Show (request)
{
mainField1 = "button with broken image";
softButtons =     (
"{\n    isHighlighted = 0;\n    softButtonID = 0;\n    systemAction = \"DEFAULT_ACTION\";\n    type = TEXT;\n}"
);
}, response: Show (response)
{
resultCode = "INVALID_DATA";
success = 0;
}, error: Error Domain=com.sdl.lifecyclemanager.error Code=-1 "INVALID_DATA" UserInfo={NSLocalizedRecoverySuggestion=Have you tried turning it off and on again?, NSLocalizedDescription=INVALID_DATA, NSLocalizedFailureReason=}
```

Reply:
The 1st update you have there is a Text and Graphic Manager update, the second a soft button update. The first has nothing to do with soft buttons.

Reply:
Yes, the first update is correct. Why is the button being sent on the second update? I am sending the exact same request twice.

Reply:
Fixed